File: app/external/toss_invest.py
import re
import threading
import time
from datetime import datetime, timedelta
from decimal import Decimal

import httpx

from app.config import get_toss_credentials
from app.schemas.system import MarketQuoteResponse, MarketSearchItem
from app.utils import to_decimal

MARKET_SUFFIX = {
    "KOSPI": "KS",
    "KOSDAQ": "KQ",
    "KONEX": "KQ",
}

_token_cache: dict[str, object] = {
    "token": None,
    "expires_at": datetime.min,
    "lock": threading.Lock(),
}


class TossApiError(ValueError):
    pass


class TossRateLimitError(TossApiError):
    pass


def is_toss_configured() -> bool:
    client_id, client_secret = get_toss_credentials()
    return bool(client_id and client_secret)


KOREAN_SYMBOL_RE = re.compile(r"^([A-Z0-9]{6})(?:\.(KS|KQ))?$")


def extract_korean_code(symbol: str) -> str | None:
    match = KOREAN_SYMBOL_RE.match(symbol.strip().upper())
    return match.group(1) if match else None


def to_toss_symbol(symbol: str) -> str:
    code = extract_korean_code(symbol)
    if code:
        return code
    return symbol.strip().upper().split(".")[0]


def to_app_symbol(toss_symbol: str, market: str | None = None) -> str:
    if re.fullmatch(r"[A-Z0-9]{6}", toss_symbol) and market:
        suffix = MARKET_SUFFIX.get(market.upper())
        if suffix:
            return f"{toss_symbol}.{suffix}"
    return toss_symbol


def _base_url() -> str:
    from app.config import get_settings

    return get_settings().toss_base_url.rstrip("/")


def _parse_api_payload(payload: dict) -> object:
    if "result" in payload:
        return payload["result"]
    if any(key in payload for key in ("items", "stocks", "prices", "data")):
        return payload
    if payload.get("error"):
        error = payload["error"]
        if isinstance(error, dict):
            message = error.get("message") or "토스증권 API 오류"
            code = error.get("code")
            if code:
                message = f"{message} ({code})"
        else:
            message = str(error)
        raise TossApiError(message)
    raise TossApiError("토스증권 API 응답 형식 오류")


def _parse_oauth_token(payload: dict) -> tuple[str, int]:
    if payload.get("error"):
        description = payload.get("error_description") or payload.get("error")
        raise TossApiError(f"토스증권 인증 실패: {description}")

    candidates: list[dict] = [payload]
    result = payload.get("result")
    if isinstance(result, dict):
        candidates.append(result)

    for data in candidates:
        token = data.get("access_token") or data.get("accessToken")
        if token:
            expires_in = data.get("expires_in") or data.get("expiresIn") or 86400
            return str(token), int(expires_in)

    raise TossApiError("토스증권 액세스 토큰을 받지 못했습니다.")


def _request_token(client_id: str, client_secret: str) -> tuple[str, datetime]:
    url = f"{_base_url()}/oauth2/token"
    with httpx.Client(timeout=15.0) as client:
        response = client.post(
            url,
            data={
                "grant_type": "client_credentials",
                "client_id": client_id,
                "client_secret": client_secret,
            },
            headers={"Content-Type": "application/x-www-form-urlencoded"},
        )
    if response.status_code == 429:
        raise TossRateLimitError("토스증권 API 요청 한도 초과")
    if response.status_code != 200:
        raise TossApiError("토스증권 인증 실패. Client ID/Secret을 확인하세요.")

    token, expires_in = _parse_oauth_token(response.json())
    return token, datetime.utcnow() + timedelta(seconds=max(expires_in - 60, 60))


def _get_access_token() -> str:
    client_id, client_secret = get_toss_credentials()
    if not client_id or not client_secret:
        raise TossApiError("토스증권 API가 설정되지 않았습니다.")

    with _token_cache["lock"]:
        token = _token_cache["token"]
        expires_at = _token_cache["expires_at"]
        if isinstance(token, str) and datetime.utcnow() < expires_at:
            return token
        token, expires_at = _request_token(client_id, client_secret)
        _token_cache["token"] = token
        _token_cache["expires_at"] = expires_at
        return token


def _api_get(path: str, params: dict | None = None) -> object:
    token = _get_access_token()
    url = f"{_base_url()}{path}"
    with httpx.Client(timeout=15.0) as client:
        response = client.get(
            url,
            params=params,
            headers={"Authorization": f"Bearer {token}"},
        )

    if response.status_code == 401:
        with _token_cache["lock"]:
            _token_cache["token"] = None
            _token_cache["expires_at"] = datetime.min
        token = _get_access_token()
        with httpx.Client(timeout=15.0) as client:
            response = client.get(
                url,
                params=params,
                headers={"Authorization": f"Bearer {token}"},
            )

    if response.status_code == 429:
        raise TossRateLimitError("토스증권 API 요청 한도 초과")
    if response.status_code != 200:
        raise TossApiError(f"토스증권 API 호출 실패 (HTTP {response.status_code})")

    return _parse_api_payload(response.json())


def _as_list(result: object) -> list[dict]:
    if isinstance(result, list):
        return [item for item in result if isinstance(item, dict)]
    if isinstance(result, dict):
        for key in ("items", "stocks", "prices", "data"):
            value = result.get(key)
            if isinstance(value, list):
                return [item for item in value if isinstance(item, dict)]
        if result and all(isinstance(value, dict) for value in result.values()):
            items: list[dict] = []
            for symbol, data in result.items():
                item = dict(data)
                item.setdefault("symbol", symbol)
                items.append(item)
            return items
        return [result]
    return []


def get_stocks(symbols: list[str]) -> list[dict]:
    if not symbols:
        return []
    result = _api_get("/api/v1/stocks", params={"symbols": ",".join(symbols)})
    return _as_list(result)


def get_prices(symbols: list[str]) -> list[dict]:
    if not symbols:
        return []
    result = _api_get("/api/v1/prices", params={"symbols": ",".join(symbols)})
    return _as_list(result)


def _stock_name(stock: dict) -> str | None:
    return stock.get("name") or stock.get("stockName")


def _stock_market(stock: dict) -> str | None:
    return stock.get("market") or stock.get("marketCode")


def _price_value(price: dict) -> Decimal | None:
    for key in ("lastPrice", "last_price", "price", "closePrice"):
        value = price.get(key)
        if value is not None:
            return to_decimal(value)
    return None


def lookup_symbol(symbol: str) -> MarketSearchItem | None:
    toss_symbol = to_toss_symbol(symbol)
    stocks = get_stocks([toss_symbol])
    if not stocks:
        return None
    stock = stocks[0]
    name = _stock_name(stock)
    if not name:
        return None
    app_symbol = to_app_symbol(stock.get("symbol") or toss_symbol, _stock_market(stock))
    return MarketSearchItem(symbol=app_symbol, name=name)


def get_quote(symbol: str) -> MarketQuoteResponse:
    toss_symbol = to_toss_symbol(symbol)
    stocks = get_stocks([toss_symbol])
    prices = get_prices([toss_symbol])

    stock = stocks[0] if stocks else {}
    price_row = prices[0] if prices else {}
    price = _price_value(price_row)
    if price is None:
        raise TossApiError("시세 조회 실패")

    app_symbol = to_app_symbol(price_row.get("symbol") or toss_symbol, _stock_market(stock))
    currency = price_row.get("currency") or stock.get("currency") or "KRW"
    return MarketQuoteResponse(
        symbol=app_symbol,
        name=_stock_name(stock),
        price=price,
        currency=currency,
        updated_at=datetime.utcnow(),
    )


def search_symbols(query: str) -> list[MarketSearchItem]:
    cleaned = query.strip().upper().rstrip(".")
    match = KOREAN_SYMBOL_RE.match(cleaned)
    if match:
        symbol = match.group(1)
    else:
        symbol = cleaned.split(".")[0]
    if not symbol or not re.fullmatch(r"[A-Z0-9.\-]+", symbol):
        return []

    item = lookup_symbol(symbol)
    return [item] if item else []


def verify_connection() -> None:
    get_prices(["005930"])


def clear_token_cache() -> None:
    with _token_cache["lock"]:
        _token_cache["token"] = None
        _token_cache["expires_at"] = datetime.min

File: src/lib/external/toss-invest.ts
import Decimal from "decimal.js";

import { getTossCredentials } from "@/lib/config";
import { toDecimal } from "@/lib/decimal";
import type { MarketQuoteResponse, MarketSearchItem } from "@/lib/validations/system";

const MARKET_SUFFIX: Record<string, string> = {
  KOSPI: "KS",
  KOSDAQ: "KQ",
  KONEX: "KQ",
};

const KOREAN_SYMBOL_RE = /^([A-Z0-9]{6})(?:\.(KS|KQ))?$/;

const tokenCache = {
  token: null as string | null,
  expiresAt: new Date(0),
  lock: Promise.resolve(),
};

export class TossApiError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "TossApiError";
  }
}

export class TossRateLimitError extends TossApiError {
  constructor(message: string) {
    super(message);
    this.name = "TossRateLimitError";
  }
}

export function isTossConfigured(): boolean {
  return !!(process.env.TOSS_CLIENT_ID && process.env.TOSS_CLIENT_SECRET);
}

export async function isTossConfiguredAsync(): Promise<boolean> {
  const credentials = await getTossCredentials();
  return !!(credentials.clientId && credentials.clientSecret);
}

export function extractKoreanCode(symbol: string): string | null {
  const match = KOREAN_SYMBOL_RE.exec(symbol.trim().toUpperCase());
  return match ? match[1] : null;
}

export function toTossSymbol(symbol: string): string {
  const code = extractKoreanCode(symbol);
  if (code) {
    return code;
  }
  return symbol.trim().toUpperCase().split(".")[0];
}

export function toAppSymbol(tossSymbol: string, market?: string | null): string {
  if (/^[A-Z0-9]{6}$/.test(tossSymbol) && market) {
    const suffix = MARKET_SUFFIX[market.toUpperCase()];
    if (suffix) {
      return `${tossSymbol}.${suffix}`;
    }
  }
  return tossSymbol;
}

async function withTokenLock<T>(fn: () => Promise<T>): Promise<T> {
  const current = tokenCache.lock;
  let release!: () => void;
  tokenCache.lock = new Promise<void>((resolve) => {
    release = resolve;
  });
  await current;
  try {
    return await fn();
  } finally {
    release();
  }
}

async function baseUrl(): Promise<string> {
  const credentials = await getTossCredentials();
  return credentials.baseUrl.replace(/\/$/, "");
}

function parseApiPayload(payload: Record<string, unknown>): unknown {
  if ("result" in payload) {
    return payload.result;
  }
  if (["items", "stocks", "prices", "data"].some((key) => key in payload)) {
    return payload;
  }
  if (payload.error) {
    const error = payload.error;
    if (typeof error === "object" && error !== null) {
      const errorObj = error as Record<string, unknown>;
      let message = String(errorObj.message ?? "토스증권 API 오류");
      const code = errorObj.code;
      if (code) {
        message = `${message} (${String(code)})`;
      }
      throw new TossApiError(message);
    }
    throw new TossApiError(String(error));
  }
  throw new TossApiError("토스증권 API 응답 형식 오류");
}

export function parseOauthToken(payload: Record<string, unknown>): [string, number] {
  if (payload.error) {
    const description = payload.error_description ?? payload.error;
    throw new TossApiError(`토스증권 인증 실패: ${String(description)}`);
  }

  const candidates: Record<string, unknown>[] = [payload];
  const result = payload.result;
  if (typeof result === "object" && result !== null) {
    candidates.push(result as Record<string, unknown>);
  }

  for (const data of candidates) {
    const token = data.access_token ?? data.accessToken;
    if (token) {
      const expiresIn = data.expires_in ?? data.expiresIn ?? 86400;
      return [String(token), Number(expiresIn)];
    }
  }

  throw new TossApiError("토스증권 액세스 토큰을 받지 못했습니다.");
}

async function requestToken(clientId: string, clientSecret: string): Promise<[string, Date]> {
  const url = `${await baseUrl()}/oauth2/token`;
  const response = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      grant_type: "client_credentials",
      client_id: clientId,
      client_secret: clientSecret,
    }),
  });

  if (response.status === 429) {
    throw new TossRateLimitError("토스증권 API 요청 한도 초과");
  }
  if (response.status !== 200) {
    throw new TossApiError("토스증권 인증 실패. Client ID/Secret을 확인하세요.");
  }

  const [token, expiresIn] = parseOauthToken((await response.json()) as Record<string, unknown>);
  const expiresAt = new Date(Date.now() + Math.max(expiresIn - 60, 60) * 1000);
  return [token, expiresAt];
}

export async function getAccessToken(): Promise<string> {
  const credentials = await getTossCredentials();
  if (!credentials.clientId || !credentials.clientSecret) {
    throw new TossApiError("토스증권 API가 설정되지 않았습니다.");
  }

  return withTokenLock(async () => {
    if (tokenCache.token && new Date() < tokenCache.expiresAt) {
      return tokenCache.token;
    }

    const [token, expiresAt] = await requestToken(credentials.clientId!, credentials.clientSecret!);
    tokenCache.token = token;
    tokenCache.expiresAt = expiresAt;
    return token;
  });
}

function asList(result: unknown): Record<string, unknown>[] {
  if (Array.isArray(result)) {
    return result.filter((item): item is Record<string, unknown> => typeof item === "object" && item !== null);
  }
  if (typeof result === "object" && result !== null) {
    const record = result as Record<string, unknown>;
    for (const key of ["items", "stocks", "prices", "data"]) {
      const value = record[key];
      if (Array.isArray(value)) {
        return value.filter((item): item is Record<string, unknown> => typeof item === "object" && item !== null);
      }
    }
    if (Object.values(record).every((value) => typeof value === "object" && value !== null)) {
      const items: Record<string, unknown>[] = [];
      for (const [symbol, data] of Object.entries(record)) {
        items.push({ ...(data as Record<string, unknown>), symbol });
      }
      return items;
    }
    return [record];
  }
  return [];
}

async function apiGet(path: string, params?: Record<string, string>): Promise<unknown> {
  const url = new URL(`${await baseUrl()}${path}`);
  if (params) {
    for (const [key, value] of Object.entries(params)) {
      url.searchParams.set(key, value);
    }
  }

  const request = async (token: string) =>
    fetch(url, {
      headers: { Authorization: `Bearer ${token}` },
    });

  let token = await getAccessToken();
  let response = await request(token);

  if (response.status === 401) {
    await withTokenLock(async () => {
      tokenCache.token = null;
      tokenCache.expiresAt = new Date(0);
    });
    token = await getAccessToken();
    response = await request(token);
  }

  if (response.status === 429) {
    throw new TossRateLimitError("토스증권 API 요청 한도 초과");
  }
  if (response.status !== 200) {
    throw new TossApiError(`토스증권 API 호출 실패 (HTTP ${response.status})`);
  }

  return parseApiPayload((await response.json()) as Record<string, unknown>);
}

async function getStocks(symbols: string[]): Promise<Record<string, unknown>[]> {
  if (symbols.length === 0) {
    return [];
  }
  const result = await apiGet("/api/v1/stocks", { symbols: symbols.join(",") });
  return asList(result);
}

async function getPrices(symbols: string[]): Promise<Record<string, unknown>[]> {
  if (symbols.length === 0) {
    return [];
  }
  const result = await apiGet("/api/v1/prices", { symbols: symbols.join(",") });
  return asList(result);
}

function stockName(stock: Record<string, unknown>): string | null {
  const name = stock.name ?? stock.stockName;
  return name != null ? String(name) : null;
}

function stockMarket(stock: Record<string, unknown>): string | null {
  const market = stock.market ?? stock.marketCode;
  return market != null ? String(market) : null;
}

function priceValue(price: Record<string, unknown>): Decimal | null {
  for (const key of ["lastPrice", "last_price", "price", "closePrice"]) {
    const value = price[key];
    if (value != null && (typeof value === "string" || typeof value === "number" || value instanceof Decimal)) {
      return toDecimal(value);
    }
  }
  return null;
}

export function parseTossPriceTimestamp(value: unknown): Date | null {
  if (value == null) {
    return null;
  }
  const parsed = new Date(String(value));
  return Number.isNaN(parsed.getTime()) ? null : parsed;
}

async function lookupSymbol(symbol: string): Promise<MarketSearchItem | null> {
  const tossSymbol = toTossSymbol(symbol);
  const stocks = await getStocks([tossSymbol]);
  if (stocks.length === 0) {
    return null;
  }
  const stock = stocks[0];
  const name = stockName(stock);
  if (!name) {
    return null;
  }
  const appSymbol = toAppSymbol(String(stock.symbol ?? tossSymbol), stockMarket(stock));
  return { symbol: appSymbol, name };
}

export async function getQuote(symbol: string): Promise<MarketQuoteResponse> {
  const tossSymbol = toTossSymbol(symbol);
  const stocks = await getStocks([tossSymbol]);
  const prices = await getPrices([tossSymbol]);

  const stock = stocks[0] ?? {};
  const priceRow = prices[0] ?? {};
  const price = priceValue(priceRow);
  if (price == null) {
    throw new TossApiError("시세 조회 실패");
  }

  const appSymbol = toAppSymbol(String(priceRow.symbol ?? tossSymbol), stockMarket(stock));
  const currency = String(priceRow.currency ?? stock.currency ?? "KRW");
  const updatedAt =
    parseTossPriceTimestamp(priceRow.timestamp ?? priceRow.tradedAt ?? priceRow.traded_at) ?? new Date();
  return {
    symbol: appSymbol,
    name: stockName(stock),
    price,
    currency,
    updated_at: updatedAt,
  };
}

export async function searchStocks(query: string): Promise<MarketSearchItem[]> {
  const cleaned = query.trim().toUpperCase().replace(/\.$/, "");
  const match = KOREAN_SYMBOL_RE.exec(cleaned);
  const symbol = match ? match[1] : cleaned.split(".")[0];
  if (!symbol || !/^[A-Z0-9.\-]+$/.test(symbol)) {
    return [];
  }

  const item = await lookupSymbol(symbol);
  return item ? [item] : [];
}

export async function verifyConnection(): Promise<void> {
  await getPrices(["005930"]);
}

export async function clearTokenCache(): Promise<void> {
  await withTokenLock(async () => {
    tokenCache.token = null;
    tokenCache.expiresAt = new Date(0);
  });
}
